File: backend/lib/L2_data/schema/goals/smartable.py
#  Copyright (c) 2022. Alexandr Moroz

from abc import ABC
from datetime import date, datetime
from typing import Optional

from ..base_schema import PKGetable, PKUpsertable, Titleable
from .person import PersonSchemaGet


class _Smartable(Titleable, ABC):
    description: Optional[str]
    due_date: Optional[datetime | date]
    parent_id: Optional[int]
    closed: bool


class SmartableGet(_Smartable, PKGetable, ABC):
    assignee: Optional[PersonSchemaGet]
    author: Optional[PersonSchemaGet]
    remote_tracker_id: Optional[int]


class SmartableUpsert(_Smartable, PKUpsertable, ABC):
    assignee_id: Optional[int]
    author_id: Optional[int]
    status_id: Optional[int]

File: frontend/lib/L1_domain/api_schema/smartable.dart
// Copyright (c) 2022. Alexandr Moroz

import 'base_upsert_schema.dart';

class SmartUpsert extends StatusableUpsert {
  SmartUpsert({
    required int? id,
    required String title,
    required bool closed,
    required this.parentId,
    required this.description,
    required this.dueDate,
    required this.statusId,
  }) : super(id: id, title: title, closed: closed);

  final int? parentId;
  final String description;
  final DateTime? dueDate;
  final int? statusId;
}
